pkg/controller/sync/aggregator/aggregator_test.go

```diff
@@ -0,0 +1,237 @@
+package aggregator
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type upsertKeyGVKs struct {
+	key  Key
+	gvks []schema.GroupVersionKind
+}
+
+func Test_bidiGVKAggregator_UpsertWithValidation(t *testing.T) {
+	// Define test cases with inputs and expected outputs
+	tests := []struct {
+		name string
+		// each entry in the list is a new Upsert call
+		keyGVKs     []upsertKeyGVKs
+		expectAdded bool
+		expectData  map[Key]map[schema.GroupVersionKind]struct{}
+		expectRev   map[schema.GroupVersionKind]map[Key]struct{}
+		expectErr   bool
```

Comment:
It can be nice create a separate Test function for error path cases, making both the happy path and error path tests simpler.

Reply:
Let's revisit testing improvements once the shape is set for aggregator